File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        rows, cols = len(heights), len(heights[0])
        visited = [[False for _ in range(cols)] for _ in range(rows)]        
        directions = [(1, 0), (0, 1), (0, -1), (-1, 0)]
        res = []

        def bfs(r, c):
            q = deque([(r, c)])
            visited[r][c] = True

            while q:
                cr, cc = q.popleft()
                if pacific_visited[cr][cc]:
                    res.append((cr,cc))
                for dr, dc in directions:
                    nr, nc = cr + dr, cc + dc
                    if (
                        0 <= nr < rows
                        and 0 <= nc < cols
                        and not visited[nr][nc]
                        and heights[nr][nc] >= heights[cr][cc]
                    ):
                        visited[nr][nc] = True
                        q.append((nr, nc))

            return False

        # check pacific side
        pacific_visited = [[False for _ in range(cols)] for _ in range(rows)]
        for c in range(0, cols):
            if not visited[0][c]:
                bfs(0, c)

        for r in range(0, rows):
            if not visited[r][0]:
                bfs(r, 0)

        pacific_visited = [row[:] for row in visited]
        visited = [[False for _ in range(cols)] for _ in range(rows)]
        # check atlantic side
        for c in range(0, cols):
            if not visited[rows - 1][c]:
                bfs(rows - 1, c)

        for r in range(0, rows):
            if not visited[r][cols - 1]:
                bfs(r, cols - 1)

        return res

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    private static final int[][] DIRECTIONS = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } };

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        int ROWS = heights.length, COLS = heights[0].length;

        // pacific side
        boolean[][] pacificVisited = new boolean[ROWS][COLS];
        for (int col = 0; col < COLS; col++)
            dfs(heights, 0, col, pacificVisited, heights[0][col]);
        for (int row = 0; row < ROWS; row++)
            dfs(heights, row, 0, pacificVisited, heights[row][0]);

        // atlantic side
        boolean[][] atlanticVisited = new boolean[ROWS][COLS];
        for (int col = 0; col < COLS; col++)
            dfs(heights, ROWS - 1, col, atlanticVisited, heights[ROWS - 1][col]);
        for (int row = 0; row < ROWS; row++)
            dfs(heights, row, COLS - 1, atlanticVisited, heights[row][COLS - 1]);

        // check common nodes
        List<List<Integer>> res = new ArrayList<>();
        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (pacificVisited[r][c] && atlanticVisited[r][c]){
                    List<Integer> commonPair = new ArrayList<>();
                    commonPair.add(r);
                    commonPair.add(c);
                    res.add(commonPair);
                }
            }
        }
        return res;
    }

    public void dfs(int[][] heights, int row, int col, boolean[][] visited, int prevHeight) {
        if (row < 0 || col < 0 ||
                row >= heights.length || col >= heights[0].length ||
                visited[row][col] ||
                heights[row][col] < prevHeight) {
            return;
        }

        visited[row][col] = true;

        for (int[] dir : DIRECTIONS) {
            dfs(heights, row + dir[0], col + dir[1], visited, heights[row][col]);
        }
    }
}